### PR TITLE
Effective time flag

### PR DESCRIPTION
**Pull request by simonbaird**
Add a new flag, `--effective-time` and use its value instead of the current time when evaluating policies.

[HACBS-1219](https://issues.redhat.com/browse/HACBS-1219)

### CODE REVIEW

**Comment by simonbaird** [comment]
I don't love how there are 300 one line diffs in 200 files, so I'm thinking about how to improve that situation.

**Comment by simonbaird** [comment]
Rebased and also changed the design quite a bit. Still doesn't have test coverage, so still in draft.

**Comment by simonbaird** [comment]
When adding the acceptance test I realized the patch doesn't pass the effective time in to the `Evaluate` function in `internal/evaluator/conftest_evaluator.go`. I had been just adding it to config.json and assuming that was enough, which is no longer true.

So there's more work to do and this will stay in draft a little longer.

File: features/validate_image.feature [hunk 1]
```diff
@@ -114,7 +114,79 @@ Feature: evaluate enterprise contract
...
+            }
+          ],
+          "success": true,
+          "signatures": ${ATTESTATION_SIGNATURES_JSON}
+        }
+      ]
+    }
+    """
+
   @focus
```
Comment (lcarva) [inline features/validate_image.feature:153]: Should this be removed?

File: internal/evaluator/conftest_evaluator.go [hunk 2]
```diff
@@ -143,34 +146,39 @@ func createConfigJSON(fs afero.Fs, dataDir string, p *policy.Policy) error {
...
 		}
 	}
-	// Check to see that we've actually added any values to the policyConfig struct.
-	// If so, we'll update the config map. Otherwise, this is skipped.
-	if (policyConfig{} != *pc) {
-		config["config"] = map[string]interface{}{
-			"policy": pc,
-		}
+
+	// Now that the future deny logic is handled in the ec-cli and not in rego,
```
Comment (lcarva) [inline internal/evaluator/conftest_evaluator.go:173]: This. I really like the fact that the time we're evaluating against does not change between each image being validated.

File: internal/policy/policy_test.go [hunk 3]
```diff
@@ -55,6 +56,9 @@ func (c *FakeKubernetesClient) FetchEnterpriseContractPolicy(ctx context.Context
 }
 
 func TestNewPolicy(t *testing.T) {
+	timeNowStr := "2022-11-23T16:30:00Z"
+	timeNow, _ := time.Parse(time.RFC3339, timeNowStr)
```
Comment (lcarva) [inline internal/policy/policy_test.go:60]: For our own future sanity, can we fail the test if `time.Parse` does return an error?

**Review by lcarva — APPROVED** [review]

File: internal/acceptance/cli/cli.go [hunk 4]
```diff
@@ -147,6 +147,9 @@ func ecCommandIsRunWith(ctx context.Context, parameters string) (context.Context
 		logger.Log(err)
 	}
 
+	logger.Logf("Stdout: %v", stdout.String())
+	logger.Logf("Stderr: %v", stderr.String())
+
```
Comment (zregvart) [inline internal/acceptance/cli/cli.go:152]: There is already `logOutput` that could be used instead, it prints much more. Having this here the stdout/stderr will be printed twice on unexpected exit status, might be a bit confusing

**Review by zregvart — APPROVED** [review]

**Comment by zregvart** [comment]
Let's merge this and address the minor issues in a followup PR

File: internal/evaluator/conftest_evaluator.go (hunk 2 above)
Comment (simonbaird) [reply internal/evaluator/conftest_evaluator.go:173]: +1, I can imagine some potential weirdness if one conftest invocation ran a millisecond before midnight and the next ran a millisecond after.

File: internal/acceptance/cli/cli.go (hunk 4 above)
Comment (simonbaird) [reply internal/acceptance/cli/cli.go:152]: Looks like this has been dealt with via ed0db955c00d0c51c0f1eb4ec96b584ff0aa2fec .

File: features/validate_image.feature (hunk 1 above)
Comment (simonbaird) [reply features/validate_image.feature:153]: Done in e3688ebdc7f68dab00e9859b9a21ff4d9555a2a0

File: internal/policy/policy_test.go (hunk 3 above)
Comment (simonbaird) [reply internal/policy/policy_test.go:60]: 8e42aa0b6b4e1aa55dab15dbdfdb63b822399a04

**Comment by simonbaird** [comment]
Thanks for all the fixups :sweat_smile:

**Pull request merged**